### PR TITLE
crimson/onode-staged-tree: fix match_stage_t related compile issue

### PR DESCRIPTION
**Pull request by cyx1231st**
Change the type to int8_t for potential negative value when
substantiating the templates.

Specifically, clang will substantiate `staged_result_t<NODE_TYPE, STAGE-1>` when `STAGE` is 0, and result in compile error.
```
../src/crimson/os/seastore/onode_manager/staged-fltree/stages/stage_types.h:355:56: error: non-type template argument evaluates to -1, which cannot be narrowed to type 'crimson::os::seastore::onode::match_stage_t' (aka 'unsigned char') [-Wc++11-narrowing]
        size_t index, const staged_result_t<NODE_TYPE, STAGE-1>& nxt_stage_result) {
                                                       ^
../src/crimson/os/seastore/onode_manager/staged-fltree/stages/stage.h:966:29: note: in instantiation of template class 'crimson::os::seastore::onode::staged_result_t<crimson::os::seastore::onode::node_type_t::INTERNAL, '\x00'>' requested here
          auto nxt_result = NXT_STAGE_T::template lower_bound<GET_KEY>(
                            ^
../src/crimson/os/seastore/onode_manager/staged-fltree/stages/stage.h:133:21: note: expanded from macro 'NXT_STAGE_T'
#define NXT_STAGE_T staged<next_param_t>
                    ^
../src/crimson/os/seastore/onode_manager/staged-fltree/stages/stage.h:966:51: note: in instantiation of function template specialization 'crimson::os::seastore::onode::staged<crimson::os::seastore::onode::staged_params_item_iterator<crimson::os::seastore::onode::node_type_t::INTERNAL> >::lower_bound<true>' requested here
          auto nxt_result = NXT_STAGE_T::template lower_bound<GET_KEY>(
```

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
https://github.com/ceph/ceph/blob/4f8d2871de8576a1137a82893055d103b4065b14/src/crimson/os/seastore/onode_manager/staged-fltree/stages/stage.h#L964-L972

@ronen-fr  I'm not sure why clang tries to compile this code block even though `IS_BOTTOM` is true. Does this mean clang doesn't recognize `constexpr` for `else`?

**Review by athanatos — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]

**Comment by ronen-fr** [comment]
@cyx1231st please note that the '-1' problem was fixed, but node.cc won't compile still. Main issue: lambda-capturing 'special references' (from structured-bindings). I will try to suggest line changes inline.

**Comment by ronen-fr** [comment]
> I'm not sure why clang tries to compile this code block even though `IS_BOTTOM` is true. Does this mean clang doesn't recognize `constexpr` for `else`?

Good question... I have tried various combinations around that issue, but failed.

**Comment by ronen-fr** [comment]
The other changes required are in staged-fltree/node.cc, which is not part of this PR. I will issue a separate PR to handle those.

**Comment by cyx1231st** [comment]
@ronen-fr see https://github.com/ceph/ceph/pull/38578/commits/29a3c3683de6ee2e16a3165f425b32153aedc9ae

OK, I can merge this.

**Comment by cyx1231st** [comment]
jenkins test make check

File: src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h [hunk 1]
```diff
@@ -133,13 +133,13 @@ class KVPool {
       : str_sizes{str_sizes}, onodes{onode_sizes} {
     ceph_assert(range2.first < range2.second);
     ceph_assert(range2.second - 1 <= (unsigned)std::numeric_limits<shard_t>::max());
-    ceph_assert(range2.second - 1 <= std::numeric_limits<pool_t>::max());
+    // ceph_assert(range2.second - 1 <= std::numeric_limits<pool_t>::max());
```
Comment (tchaikov) [inline src/crimson/os/seastore/onode_manager/staged-fltree/tree_utils.h:136]: better off removing dead code.

**Review by ronen-fr — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]

**Pull request merged**